src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,561 @@
+        if v.get_type_code() != TYPE_CODE_LITERAL {
+            values_len += v.binary_len() - LENGTH_TYPE;
+        }
+    }
+    // object: element-count size key-entry* value-entry* key* value*
+    LENGTH_U32 + LENGTH_U32 + key_entries_len + value_entries_len + keys_len + values_len
+}
+
+#[allow(dead_code)]
+fn get_array_binary_len(data: &[Json]) -> usize {
+    let element_count = data.len();
+    let value_entries_len = element_count * LENGTH_VALUE_ENTRY;
+    let mut values_len = 0;
+    for v in data {
+        if v.get_type_code() != TYPE_CODE_LITERAL {
+            values_len += v.binary_len() - LENGTH_TYPE;
+        }
+    }
+    // array ::= element-count size value-entry* value*
+    LENGTH_U32 + LENGTH_U32 + value_entries_len + values_len
+}
+
+#[allow(dead_code)]
+fn get_str_binary_len(data: &str) -> usize {
+    let len = data.as_bytes().len();
+    get_var_u64_blen(len as u64) + len
+}
+
+#[allow(dead_code)]
+fn get_var_u64_blen(mut v: u64) -> usize {
+    let mut len = 1;
+    while v >= 0x80 {
+        v >>= 7;
+        len += 1;
+    }
+    len
+}
+
+impl Serialize for Json {
+    fn serialize<S>(&self, serializer: S) -> SResult<S::Ok, S::Error>
+        where S: Serializer
+    {
+        match *self {
+            Json::Literal(l) => {
+                match l {
+                    JSON_LITERAL_NIL => serializer.serialize_none(),
+                    JSON_LITERAL_TRUE => serializer.serialize_bool(true),
+                    _ => serializer.serialize_bool(false),
+                }
+            }
+            Json::String(ref s) => serializer.serialize_str(s),
+            Json::Object(ref obj) => {
+                let mut map = try!(serializer.serialize_map(Some(obj.len())));
+                for (k, v) in obj {
+                    try!(map.serialize_entry(k, v));
+                }
+                map.end()
+            }
+            Json::Array(ref array) => {
+                let mut tup = try!(serializer.serialize_tuple(array.len()));
+                for item in array {
+                    try!(tup.serialize_element(item));
+                }
+                tup.end()
+            }
+            Json::Double(d) => serializer.serialize_f64(d),
+            Json::I64(d) => serializer.serialize_i64(d),
+        }
+    }
+}
+
+struct JsonVisitor;
+impl<'de> Visitor<'de> for JsonVisitor {
+    type Value = Json;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "may be any string")
+    }
+
+    fn visit_unit<E>(self) -> SResult<Self::Value, E>
+        where E: de::Error
+    {
+        Ok(Json::Literal(JSON_LITERAL_NIL))
+    }
+
+    fn visit_bool<E>(self, v: bool) -> SResult<Self::Value, E>
+        where E: de::Error
+    {
+        if v {
+            Ok(Json::Literal(JSON_LITERAL_TRUE))
+        } else {
+            Ok(Json::Literal(JSON_LITERAL_FALSE))
+        }
+    }
+
+    fn visit_i64<E>(self, v: i64) -> SResult<Self::Value, E>
```

Comment:
It's not necessary to implement since [`visit_i8`](https://docs.serde.rs/src/serde/de/mod.rs.html#1091), `visit_i16`'s implementations are based on `visit_i64`, so does `visit_f32` @tennix